### PR TITLE
[LLVM][CodeGen] Add lowering for scalable vector bfloat operations.

### PR DESCRIPTION
**Pull request by paulwalker-arm**
Specifically:
  fabs, fadd, fceil, fdiv, ffloor, fma, fmax, fmaxnm, fmin, fminnm,
  fmul, fnearbyint, fneg, frint, fround, froundeven, fsub, fsqrt &
  ftrunc

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 1]
```diff
@@ -28466,6 +28520,40 @@ SDValue AArch64TargetLowering::LowerFixedLengthInsertVectorElt(
...
+    for (const SDValue &V : Op->op_values()) {
+      LoOps.push_back(DAG.getExtractSubvector(DL, MVT::nxv4bf16, V, 0));
+      HiOps.push_back(DAG.getExtractSubvector(DL, MVT::nxv4bf16, V, 4));
+    }
+
+    unsigned Opc = Op.getOpcode();
+    SDValue SplitOpLo = DAG.getNode(Opc, DL, MVT::nxv4bf16, LoOps);
+    SDValue SplitOpHi = DAG.getNode(Opc, DL, MVT::nxv4bf16, HiOps);
+    return DAG.getNode(ISD::CONCAT_VECTORS, DL, VT, SplitOpLo, SplitOpHi);
+  }
```
Comment (arsenm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:28543]: Seems like the generic vector legalizer could have gotten this?

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 2]
```diff
@@ -28466,6 +28520,40 @@ SDValue AArch64TargetLowering::LowerFixedLengthInsertVectorElt(
...
+  // Promote to float and try again.
+  EVT PromoteVT = VT.changeVectorElementType(MVT::f32);
+
+  SmallVector<SDValue, 4> Ops;
+  for (const SDValue &V : Op->op_values())
+    Ops.push_back(DAG.getNode(ISD::FP_EXTEND, DL, PromoteVT, V));
+
+  SDValue PromotedOp = DAG.getNode(Op.getOpcode(), DL, PromoteVT, Ops);
+  return DAG.getNode(ISD::FP_ROUND, DL, VT, PromotedOp,
+                     DAG.getIntPtrConstant(0, DL, true));
```
Comment (arsenm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:28554]: This is very generic and I would hope doesn't need repeating in target code

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 1 above)
Comment (paulwalker-arm) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:28543]: This was just me not looking for trouble :)

I've got a target neutral implementation but as you'd expect the effect is more widespread (all positive so far), so I'll pull it into a separate PR and then update this one if/when it lands.

Comment (paulwalker-arm) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: I was perhaps being too adventurous but by restricting the expansion to bf16 vectors the fallout shrank significantly so I've kept it with this PR to maintain sufficient testing.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp [hunk 3]
```diff
@@ -1070,13 +1071,21 @@ void VectorLegalizer::Expand(SDNode *Node, SmallVectorImpl<SDValue> &Results) {
     break;
   case ISD::FMINNUM:
   case ISD::FMAXNUM:
+    if (SDValue Expanded = ExpandBF16Arith(Node)) {
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp:1074]: Probably shouldn't just bypass the dedicated function to lower something

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp [hunk 4]
```diff
@@ -2134,6 +2166,67 @@ bool VectorLegalizer::tryExpandVecMathCall(
   return tryExpandVecMathCall(Node, LC, Results);
 }
 
+// Try to lower BFloat arithmetic by performing the same operation on operands
+// that have been promoted to Float32, the result of which is then truncated.
+// If promotion requires non-legal types the operation is split with the
+// promotion occuring during a successive call to this function.
+SDValue VectorLegalizer::ExpandBF16Arith(SDNode *Node) {
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp:2173]: This is just the promote action? It shouldn't be getting used for expand?

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp (hunk 3 above)
Comment (paulwalker-arm) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp:1074]: I was a little unsure how generic `expandFMINNUM_FMAXNUM` actually is.  The description seemed pretty specific alongside a use in SIISelLowering, so wasn't sure if I could just add the splitting code into it.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp (hunk 4 above)
Comment (paulwalker-arm) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp]: I've removed all the promotion code and am instead making use of common code via setOperationPromotedToType.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp (hunk 3 above)
Comment (paulwalker-arm) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp]: I've moved the expansion function into TargetLowering and updated the two expandFMIN###_FMAXNUM### functions to make us of it.

File: llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td [hunk 5]
```diff
@@ -663,6 +663,13 @@ let Predicates = [HasSVEorSME] in {
   defm FABS_ZPmZ : sve_int_un_pred_arit_1_fp<0b100, "fabs", AArch64fabs_mt>;
   defm FNEG_ZPmZ : sve_int_un_pred_arit_1_fp<0b101, "fneg", AArch64fneg_mt>;
 
+  foreach VT = [nxv2bf16, nxv4bf16, nxv8bf16] in {
```
Comment (huntergr-arm) [inline llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td:666]: nit: May be worth a comment clarifying that you're zeroing or inverting the sign bit.

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 6]
```diff
@@ -11917,3 +11923,38 @@ bool TargetLowering::LegalizeSetCCCondCode(SelectionDAG &DAG, EVT VT,
...
+SDValue TargetLowering::expandVectorNaryOpBySplitting(SDNode *Node,
+                                                      SelectionDAG &DAG) const {
+  EVT VT = Node->getValueType(0);
+  // Despite its documentation, GetSplitDestVTs will assert if VT cannot be
+  // split into two equal parts.
+  if (!VT.isVector() || !VT.getVectorElementCount().isKnownMultipleOf(2))
+    return SDValue();
+
+  EVT LoVT, HiVT;
+  std::tie(LoVT, HiVT) = DAG.GetSplitDestVTs(VT);
```
Comment (huntergr-arm) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:11936]: Not a request, more of an style option to match the use of auto elsewhere in the patch for you to consider.
```suggestion
  auto [LoVT, HiVT] = DAG.GetSplitDestVTs(VT);
```
That lets you get rid of the separate declaration line.

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 7]
```diff
@@ -11917,3 +11923,38 @@ bool TargetLowering::LegalizeSetCCCondCode(SelectionDAG &DAG, EVT VT,
...
+  unsigned Opcode = Node->getOpcode();
+
+  // Don't expand if the result is likely to be unrolled anyway.
+  if (!isOperationLegalOrCustomOrPromote(Opcode, LoVT))
+    return SDValue();
+
+  SmallVector<SDValue, 4> LoOps, HiOps;
+  for (const SDValue &V : Node->op_values()) {
+    SDValue Lo, Hi;
+    std::tie(Lo, Hi) = DAG.SplitVector(V, DL, LoVT, HiVT);
```
Comment (huntergr-arm) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:11952]: Same structured binding style can be applied here.

**Review by huntergr-arm — APPROVED** [review]
LGTM

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp (hunk 6 above)
Comment (paulwalker-arm) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp]: Oh nice. Done.

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp (hunk 7 above)
Comment (paulwalker-arm) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp]: Done.

File: llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td (hunk 5 above)
Comment (paulwalker-arm) [reply llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td:666]: Done.

**Pull request merged**